### PR TITLE
AVRO-3478: Add Duration logical type to C#

### PR DESCRIPTION
**Pull request by zcsizmadia**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Avro Jira](https://issues.apache.org/jira/browse/AVRO-3478
  - https://issues.apache.org/jira/browse/AVRO-3478
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](https://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

**Comment by KyleSchoonover** [comment]
The months part throws me off in the spec.  There isn't a default month size.  Did any other language implement this?  I couldn't find it in Java.

**Comment by zcsizmadia** [comment]
I know rust has it. I just addedd it for completeness. @martin-g might be able to help me out there. I am not usre where I saw that the days in month is 30. However rust impleent it a simple tuple <int,int,int>

**Comment by zcsizmadia** [comment]
After reading the rust code more carefully, the C# implementation should be most likely by adding an AvroDuration struct, which is a months, days, milliseconds tupple. Just like the AVroDecimal is implemented

**Comment by zcsizmadia** [comment]
Instead of using TimeSpan base object for the Duration logical type, it uses a new `AvruDuration` struct, similarly rust implements the duration logical type.

**Comment by KyleSchoonover** [comment]
My conflict is more with the spec.  I would expect to see more days, hours, minutes, seconds, milliseconds.

**Comment by zcsizmadia** [comment]
Avro spec: https://github.com/apache/avro/blob/master/doc/content/en/docs/next/Specification/_index.md#duration
Rust implementation: https://github.com/apache/avro/blob/master/lang/rust/avro/src/duration.rs

@KyleSchoonover Are you questioning the spec or the implementation?

**Comment by KyleSchoonover** [comment]
> Avro spec: https://github.com/apache/avro/blob/master/doc/content/en/docs/next/Specification/_index.md#duration
> Rust implementation: https://github.com/apache/avro/blob/master/lang/rust/avro/src/duration.rs
> 
> @KyleSchoonover Are you questioning the spec or the implementation?
> 

The spec.  I feel it should be changed to drop months as a measurement of duration.

**Comment by martin-g** [comment]
The discussion how to design it was in ticket https://issues.apache.org/jira/browse/AVRO-739.
The suggestion for `interval` (duration) is in comment: https://issues.apache.org/jira/browse/AVRO-739?focusedCommentId=14083646&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14083646

So far I didn't need to debug issues in duration.rs but looking at its code and the spec it seems the implementation follows the spec.

If you want to discuss the spec I'd suggest to do it at dev@avro.apache.org.

**Comment by martin-g** [comment]
@zcsizmadia I guess you saw `30` at https://github.com/apache/avro/pull/1263/files#diff-f676b1f77f65cebe90fecda20258df525c0cfeaca931f7327589bc75214ec6c0R154
This is a PR that adds support for duration for Java.

**Comment by martin-g** [comment]
How we will proceed here ?
@KyleSchoonover Do you approve ?

File: lang/csharp/src/apache/main/Util/LogicalType.cs [hunk 1]
```diff
@@ -74,4 +75,32 @@ public virtual void ValidateSchema(LogicalSchema schema)
         /// <param name="logicalValue">The logical value to test.</param>
         public abstract bool IsInstanceOfLogicalType(object logicalValue);
     }
+
+    /// <summary>
+    /// Base for all logical type implementations that are based on a struct type.
+    /// </summary>
+    public abstract class LogicalType<T> : LogicalType
+        where T : struct
```
Comment (KalleOlaviNiemitalo) [inline lang/csharp/src/apache/main/Util/LogicalType.cs:83]: That seems a good idea.

**Review by KalleOlaviNiemitalo — CHANGES_REQUESTED** [review]

**Pull request closed without merging**